curriculum/challenges/english/25-front-end-development/lab-color-picker/67bf4350777ac6ffdc027745.md

```diff
@@ -0,0 +1,320 @@
+4. When `#color-input` is changed, `#color-picker-container` should have its background set to that new value.
+
+# --hints--
+
+You should use the `useState` hook.
+
+```js
+assert.match(code, /useState/);
+```
+
+You should use the return value of the `useState` hook.
+
+```js
+assert.match(code, /=\s*(?:React\.)?useState\(/);
+```
+
+You should use array destructuring to assign the return value of `useState` to two variables.
+
+```js
+assert.match(code, /\[.*,.*\]\s*=\s*(?:React\.)?useState\(/);
+```
+
+You should pass `#ffffff` as the initial value to your `useState` call.
+
+```js
+assert.match(code, /useState\(\s*('|"|`)#ffffff\1\s*\)/i);
+```
+
+You should render an element with `color-picker-container` as the `id`.
+
+```js
+assert.exists(document.getElementById("color-picker-container"));
+```
+
+Your `#color-picker-container` should have a white background initially.
+
+```js
+const el = document.getElementById("color-picker-container");
+const style = window.getComputedStyle(el);
+assert.equal(style.backgroundColor, "rgb(255, 255, 255)");
+```
+
+You should render an `input` element with an `id` of `color-input`.
+
+```js
+const inputMaybe = document.getElementById("color-input");
+assert.exists(inputMaybe);
+assert.equal(inputMaybe?.tagName, "INPUT");
+```
+
+Your `#color-input` should be a `color` input.
+
+```js
+const inputDefinitely = document.getElementById("color-input");
+assert.equal(inputDefinitely?.type, "color");
+```
+
+Your `#color-input` should have an `onChange` attribute.
+
+```js
+const definitelyAnInput = document.getElementById("color-input");
+assert.exists(definitelyAnInput);
+const key = Object.keys(definitelyAnInput).find(key => key.startsWith("__reactProps"));
+const reactInstance = definitelyAnInput[key]
+assert.exists(reactInstance);
+assert.isFunction(reactInstance.onChange);
+```
+
+Your `#color-input` should have an initial value of `#ffffff`.
+
+```js
+const mostCertainlyAnInput = document.getElementById("color-input");
+assert.exists(mostCertainlyAnInput);
+assert.strictEqual(mostCertainlyAnInput.value, "#ffffff");
+```
+
+The initial value of your `#color-input` should be managed by your `useState` hook.
+
+```js
+const declaredState = code.match(/const\s*\[\s*(.*)\s*,.*\]\s*=\s*useState/);
+console.log(declaredState);
+const declaredStateVar = declaredState?.[1];
+console.log(declaredStateVar);
+const regex = new RegExp(`value\s*=\s*\{\s*${declaredStateVar}\s*}`);
+assert.match(code, regex);
+```
+
+When your `#color-input` value is changed, your `#color-picker-container` should change its background color to reflect that value.
+
+```js
+const inputForSure = document.getElementById("color-input");
+assert.exists(inputForSure);
+const nativeValueSetter = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, "value").set;
+nativeValueSetter.call(inputForSure, "#000001");
+inputForSure.dispatchEvent(new Event("input", { bubbles: true}));
```

Comment:
React complains for me, because this is not wrapped in `await React.act(async () => {})`.

Reply:
Have I mentioned that I hate React with a passion?

Reply:
```console
Warning: An update to MyApp inside a test was not wrapped in act(...).

When testing, code that causes React state updates should be wrapped into act(...):

act(() => {
  /* fire events that update state */
});
/* assert on the output */

This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
```


It seems intermittant.